static/src/javascripts/projects/commercial/modules/dfp/on-slot-viewable.ts

```diff
@@ -1,38 +1,103 @@
+import type { AdSizeString } from '@guardian/commercial-core';
+import { adSizes } from '@guardian/commercial-core';
+import config from 'lib/config';
 import { getUrlVars } from '../../../../lib/url';
+import type { Advert } from './Advert';
 import { getAdvertById } from './get-advert-by-id';
 import { enableLazyLoad } from './lazy-load';
+import { memoizedFetchNonRefreshableLineItemIds } from './non-refreshable-line-items';
 
-const setSlotAdRefresh = (event: googletag.events.Event): void => {
+const outstreamSizes = [
+	adSizes.outstreamDesktop.toString(),
+	adSizes.outstreamMobile.toString(),
+	adSizes.outstreamGoogleDesktop.toString(),
+];
+
+/**
+ * Determine whether an advert should refresh, taking into account
+ * its size, whether there's a pageskin or whether the advert's
+ * line item is marked as non-refreshable
+ *
+ *  - Fluid ads should not refresh
+ *  - Outstream ads should not refresh
+ *  - Pageskins should not refresh
+ *  - Ads that have line items marked as non-refreshable should not be
+ * 	  refreshed. This information is retrieved via the DFP non refreshable
+ * 	  line item API endpoint
+ *
+ * @param advert The candidate advert to check
+ * @param nonRefreshableLineItemIds The array of line item ids for which
+ * adverts should not refresh
+ */
+const shouldRefresh = (
+	advert: Advert,
+	nonRefreshableLineItemIds: number[] = [],
+): boolean => {
+	const sizeString = advert.size?.toString();
+	const isNotFluid = sizeString !== '0,0';
+	const isOutstream =
+		sizeString && outstreamSizes.includes(sizeString as AdSizeString);
+	const isNonRefreshableLineItem =
+		advert.lineItemId &&
+		nonRefreshableLineItemIds.includes(advert.lineItemId);
+
+	return (
+		isNotFluid &&
+		!isOutstream &&
+		!config.get('page.hasPageSkin') &&
+		!isNonRefreshableLineItem
+	);
```

Comment:
This is also code that’s identical to on-slot-render, so we might want to extract it to a helper function?

We could also use early returns, which can help understand the code paths a little bit better:

```ts
	const sizeString = advert.size?.toString();
	
	if(!sizeString) === return false; // is this right?
	
	if(sizeString === '0,0') return false;
	
	if (outstreamSizes.includes(sizeString as AdSizeString)) return false;
	
	if(advert.lineItemId &&
		nonRefreshableLineItemIds.includes(advert.lineItemId)) return false;

	return true;
```

Reply:
Good call, I've updated to reflect this.